File: pixiejsClient/src/app/network/packets/PingPacket.ts
import { EvPacketWriter } from '../EvPacketWriter';
import { EvPacketReader } from '../EvPacketReader';
import { PacketHeader } from '../PacketHeader';
import { PacketId } from '../PacketHandler';

export class PingPacket {
    header: PacketHeader
    ping: number
    time: bigint

    constructor(header: PacketHeader, ping: number, time: bigint) {
        this.header = header
        this.ping = ping
        this.time = time
    }

    static create(): ArrayBuffer {
        const writer = new EvPacketWriter(PacketId.Ping);
        writer
            .i16(0) // Ping field (ushort)
            .i64(BigInt(Date.now() * 10000)) // TickCounter (long) - convert milliseconds to .NET ticks
            .FinishPacket();
        return writer.ToArray();
    }

    static fromBuffer(buffer: ArrayBuffer): PingPacket {
        const reader = new EvPacketReader(buffer);
        const header = reader.Header();
        const ping = reader.i16();
        const time = reader.i64();
        return new PingPacket(header, ping, time);
    }

    static handle(buffer: ArrayBuffer): number {
        const packet = PingPacket.fromBuffer(buffer);
        const latency = Date.now() - Number(packet.time / 10000n); // Convert .NET ticks to milliseconds
        return latency;
    }
}

File: server/Simulation/Net/PingPacket.cs
using System;
using server.Enums;

namespace server.Simulation.Net;

public class PingPacket
{
    public ushort Ping { get; set; }
    public long TickCounter { get; set; }

    public static Memory<byte> Create()
    {
        using var writer = new PacketWriter(PacketId.Ping);
        writer.WriteUInt16(0)
        .WriteInt64(DateTime.UtcNow.Ticks);
        return writer.Finalize();
    }

    public static PingPacket Read(Memory<byte> buffer)
    {
        var reader = new PacketReader(buffer);

        return new PingPacket
        {
            Ping = reader.ReadUInt16(),
            TickCounter = reader.ReadInt64()
        };
    }
}